src/v/datalake/translation/partition_translator.cc

```diff
@@ -0,0 +1,273 @@
+      _writer_scratch_space,
+      fmt::format("rp-parquet-translator-term-{}", _term),
+      max_rows_per_row_group,
+      max_bytes_per_row_group);
+    auto multiplexer = datalake::record_multiplexer(std::move(writer_factory));
+    auto result = co_await std::move(rdr).consume(
+      std::move(multiplexer), model::no_timeout);
+    if (result.has_error()) {
+        vlog(_logger.warn, "Error translating range {}", result.error());
+        co_return std::nullopt;
+    }
+    co_return std::move(result.value());
+}
+
+ss::future<partition_translator::translation_success>
+partition_translator::do_translate_once() {
+    if (
+      !_partition->get_ntp_config().iceberg_enabled()
+      || !_features->local().is_active(features::feature::datalake_iceberg)) {
+        co_return translation_success::no;
+    }
+    auto reconcile_result = co_await reconcile_with_coordinator();
+    if (!reconcile_result) {
+        vlog(_logger.warn, "reconciliation with coordinator failed");
+        co_return translation_success::no;
+    }
+    vlog(_logger.trace, "reconciliation result: {}", reconcile_result.value());
+    auto read_begin_offset = kafka::next_offset(reconcile_result.value());
+    auto read_end_offset = max_offset_for_translation();
+    if (read_begin_offset >= read_end_offset) {
+        vlog(
+          _logger.trace,
+          "translation up to date until kafka offset: {}, lso: {}, nothing to "
+          "do",
+          _partition->last_stable_offset());
+        co_return translation_success::yes;
+    }
+    // We have some data to translate, make a reader
+    // and dispatch to the iceberg translator
+    //
+    // todo: what if we have too little to read, eg: slow moving topic.
+    // Currently we rely on user setting the translation interval
+    // appropriately to ensure each attempt has something substantial
+    // to read. We can add some heuristics to estimate the size and
+    // backoff if it is too small and wait for enough data to
+    // accumulate. The resulting parquet files are only performant
+    // if there is a big chunk of data in them. Smaller parquet files
+    // are an overhead for iceberg metadata.
+    auto proxy = kafka::make_partition_proxy(_partition);
+    auto log_reader = co_await proxy.make_reader(
+      {kafka::offset_cast(read_begin_offset),
+       kafka::offset_cast(read_end_offset),
+       0,
+       _max_bytes_per_reader,
+       datalake_priority(),
+       std::nullopt,
+       std::nullopt,
+       _as});
+
+    vlog(
+      _logger.debug,
+      "translating data in kafka range: [{}, {}]",
+      read_begin_offset,
+      read_end_offset);
+    auto tracker = kafka::aborted_transaction_tracker::create_default(
+      &proxy, std::move(log_reader.ot_state));
+    auto kafka_reader = kafka::read_committed_reader(
+      std::move(tracker), std::move(log_reader.reader));
+    auto units = co_await ss::get_units(_parallel_translations, 1, _as);
```

Comment:
maybe we can grab units a bit earlier, when we actually create a reader not to allocate any resources. Now the readear is read and waiting for the semaphore

Reply:
I can move it a few lines above but the reader is lazy right? It doesn't read anything until L175+ where we pass a consumer?